### PR TITLE
Fast-path Linq Join on empty

### PR DESCRIPTION
**Pull request by JonHanna**
If outer enumerable is empty, avoid creating `Lookup<TKey, TInner>` (c.f. #5372).

If lookup is created but it is empty, short-circuit out of method.

cc: @stephentoub, @VSadov

### CODE REVIEW

File: src/System.Linq/src/System/Linq/Join.cs [hunk 1]
```diff
@@ -31,15 +31,27 @@ public static partial class Enumerable
...
-                        yield return resultSelector(item, g.elements[i]);
+                        do
+                        {
+                            TOuter item = e.Current;
+                            Grouping<TKey, TInner> g = lookup.GetGrouping(outerKeySelector(item), false);
+                            if (g != null)
+                            {
+                                int count = g.count;
+                                TInner[] elements = g.elements;
+                                for (int i = 0; i != g.count; ++i)
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Join.cs]: ```count```?

**Comment by stephentoub** [comment]
LGTM

File: src/System.Linq/src/System/Linq/Join.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Join.cs]: Yes, I did.

**Comment by VSadov** [comment]
LGTM

**Pull request merged**